hazelcast-client/src/test/java/com/hazelcast/client/flakeidgen/impl/FlakeIdGenerator_NodeIdOverflowIntegrationTest.java

```diff
@@ -44,23 +46,31 @@
     public ExpectedException exception = ExpectedException.none();
 
     private TestHazelcastFactory factory = new TestHazelcastFactory();
+    private HazelcastInstance instance2;
+    private HazelcastInstance instance1;
+
+    @Before
+    public void before() {
+        Config cfg = new Config();
+        cfg.getFlakeIdGeneratorConfig("gen").setPrefetchCount(1);
+        instance1 = factory.newHazelcastInstance(cfg);
+        instance2 = factory.newHazelcastInstance(cfg);
+    }
 
     @After
     public void after() {
         factory.shutdownAll();
     }
 
     @Test
-    public void when_memberOutOfRangeNodeId_then_theOtherMemberUsed() throws Exception {
+    public void when_memberOutOfRangeNodeId_then_theOtherMemberUsed() {
         // Design of this test: we create two members. To one of them, out-of-range memberListJoinVersion
         // is assigned. This causes the client message to fail and the client will retry with another member
         // and will never again try the failed member.
         //
         // Probability of initially choosing a good member and not going through the process of choosing another
         // one is 50%. To avoid false success, we retry 10 times, which gives the probability of false success
         // of 0.1%.
-        HazelcastInstance instance1 = factory.newHazelcastInstance();
-        HazelcastInstance instance2 = factory.newHazelcastInstance();
         assignOverflowedNodeId(instance2);
 
         ClientConfig clientConfig = new ClientConfig();
```

Comment:
In order for this test to serve its purpose, we should alternate the address the client connects to in each iteration. Something like this before creating the client:
```
clientConfig.getNetworkConfig().setAddresses(Collections.singletonList(
                    i % 2 == 0 ? formatAddress(instance1.getLocalEndpoint())
                            : formatAddress(instance2.getLocalEndpoint())
            ));
```

```
private static String formatAddress(Endpoint endpoint) {
        InetSocketAddress socketAddress = (InetSocketAddress) endpoint.getSocketAddress();
        return socketAddress.getHostName() + ":" + socketAddress.getPort();
    }
```

Reply:
We use non-smart client, that client should choose random target and use it all the time. The problem is that it always chose the 1st instance. Sancar fixed this (not yet merged) and the test started to fail, that's how we discovered this. So the test tests what it's supposed to.

Reply:
I tried your approach, but it doesn't really choose the target the client connects to. However, it failed, always on the 4th iteration (4th i think). I don't understand this.

Reply:
> but it doesn't really choose the target the client connects to

It does, it's normal to get the failure on 4th iteration:
- 1st iteration picks `instance1`, return new IDs successfully
- 2nd iteration picks `instance2`: at this point, the proxy's `this.nodeId` is `-1` (not picked yet); `getNodeId` assigns the `memberListVersion` (that is 100000) to `this.nodeId` and returns `nodeId` (-2) --> obtains IDs from other member via operation
- 3rd iteration picks `instance1`, returns successfully
- 4th iteration: proxy's `this.nodeId` is already 100000, `newIdBaseLocal` throws `NodeIdOutOfRangeException`

Anyway, since a further PR will introduce random choice of member for non-smart client, there is no need to add this logic here.

Reply:
You're probably right...